critwall/src/main/java/org/opencadc/critwall/FileSyncJob.java

```diff
@@ -0,0 +1,212 @@
+import org.opencadc.inventory.db.ArtifactDAO;
+import org.opencadc.inventory.storage.NewArtifact;
+import org.opencadc.inventory.storage.StorageAdapter;
+
+import org.opencadc.inventory.storage.StorageMetadata;
+
+public class FileSyncJob implements Runnable {
+    private static final Logger log = Logger.getLogger(FileSyncJob.class);
+
+    private final ArtifactDAO artifactDAO;
+    private final URI artifactID;
+    private final URI resourceID;
+    private final StorageAdapter storageAdapter;
+
+    public FileSyncJob(URI artifactID, URI resourceID, StorageAdapter storageAdapter, ArtifactDAO artifactDAO) {
+        InventoryUtil.assertNotNull(FileSyncJob.class, "artifactID", artifactID);
+        InventoryUtil.assertNotNull(FileSyncJob.class, "resourceID", resourceID);
+        InventoryUtil.assertNotNull(FileSyncJob.class, "storageAdapter", storageAdapter);
+        InventoryUtil.assertNotNull(FileSyncJob.class, "artifactDAO", artifactDAO);
+
+        this.artifactID = artifactID;
+        this.resourceID = resourceID;
+        this.storageAdapter = storageAdapter;
+        this.artifactDAO = artifactDAO;
+    }
+
+    @Override
+    public void run() {
+        RegistryClient regClient = new RegistryClient();
+        log.debug("resource id: " + this.resourceID);
+        URL certURL = regClient.getServiceURL(this.resourceID, Standards.VOSPACE_SYNC_21, AuthMethod.CERT);
+        log.debug("certURL: " + certURL);
+
+        try {
+            // Ask for all protocols available back, and the server will
+            // give you URLs for the ones it allows.
+            List<Protocol> protocolList = new ArrayList<Protocol>();
+            Protocol httpsCert = new Protocol(VOS.PROTOCOL_HTTPS_GET);
+
+            log.debug("protocols: " + httpsCert);
+            protocolList.add(httpsCert);
+
+            Transfer transfer = new Transfer(artifactID, Direction.pullFromVoSpace, protocolList);
+            transfer.version = VOS.VOSPACE_21;
+
+            TransferWriter writer = new TransferWriter();
+            ByteArrayOutputStream out = new ByteArrayOutputStream();
+            writer.write(transfer, out);
+            FileContent content = new FileContent(out.toByteArray(), "text/xml");
+            log.debug("xml file content to be posted: " + transfer);
+
+            log.debug("artifact path: " + artifactID.getPath());
+            HttpPost post = new HttpPost(certURL, content, true);
+            post.prepare();
+            log.debug("post prepare done");
+
+            // TODO: error handling is likely very different than this...
+            if (post.getThrowable() != null && post.getThrowable() instanceof FileNotFoundException) {
+                throw (FileNotFoundException) post.getThrowable();
+            }
+
+            TransferReader reader = new TransferReader();
+            Transfer t = reader.read(post.getInputStream(), null);
+            List<String> urlList = t.getAllEndpoints();
+
```

Comment:
Ouch - getAllEndpoints() shows how dated and sloppy the cadc-vos lib still is :-)
I think you could move all the code above up to this point into a method, maybe
`private List<URL> getDownloadURLs(...)`
The negotiation part cannot really be retried (except for some transient failures that HttpPost already retries) so if it fails the job needs to give up and return.

Reply:
Put this code into a function, try/catch around function call to quit run early if this fails.
Decided to prune malformed URLs from the list at this point: as long as one is good, something can be tried.